### ISSUE TITLE
krb5: --enable-dns-for-realm doesn't take effect as expected

### ISSUE THREAD

**Issue by lark** [termux/termux-packages#3326]
**Problem description**

> `$ kinit randomuser@ATHENA.MIT.EDU`
> `kinit: Cannot find KDC for realm "ATHENA.MIT.EDU" while getting initial credentials`
 
kinit doesn't work as expected even if DNS records are setup correctly.

**Steps to reproduce**

1. edit /data/data/com.termux/files/usr/etc/krb5.conf
2. add these two lines in [libdefaults] section
> dns_lookup_realm = true
> dns_lookup_kdc = true
3. use kinit to initialize a principal

**Expected behavior**

Password should be asked for this principal, or principal not found error is emitted.

**Additional information**

Package: krb5
Version: 1.16.2

**Cause of issue**

`configure` script doesn't work as expected.

Here is a trace of `confdefs.h` during build

>/* confdefs.h */
#define PACKAGE_NAME "Kerberos 5"
#define PACKAGE_TARNAME "krb5"
#define PACKAGE_VERSION "1.16.2"
#define PACKAGE_STRING "Kerberos 5 1.16.2"
#define PACKAGE_BUGREPORT "krb5-bugs@mit.edu"
#define PACKAGE_URL ""
#define STDC_HEADERS 1
#define HAVE_SYS_TYPES_H 1
#define HAVE_SYS_STAT_H 1
#define HAVE_STDLIB_H 1
#define HAVE_STRING_H 1
#define HAVE_MEMORY_H 1
#define HAVE_STRINGS_H 1
#define HAVE_INTTYPES_H 1
#define HAVE_STDINT_H 1
#define HAVE_UNISTD_H 1
#define _GNU_SOURCE 1
#define __STDC_WANT_LIB_EXT1__ 1
#define CONFIG_SMALL 1
#define HAVE_COM_ERR_INTL 1
#define HAVE_PRAGMA_WEAK_REF 1
#define DELAY_INITIALIZER 1
#define CONSTRUCTOR_ATTR_WORKS 1
#define DESTRUCTOR_ATTR_WORKS 1
#define ENABLE_THREADS 1
#define HAVE_PTHREAD_PRIO_INHERIT 1
#define HAVE_PTHREAD 1
#define HAVE_PTHREAD_ONCE 1
#define HAVE_PTHREAD_RWLOCK_INIT 1
#define HAVE_PTHREAD_RWLOCK_INIT_IN_THREAD_LIB 1
#define USE_DLOPEN 1
#define HAS_VA_COPY 1
#define CAN_COPY_VA_LIST 1
#define HAVE_REGCOMP 1
#define GETSOCKNAME_ARG3_TYPE socklen_t
#define GETPEERNAME_ARG3_TYPE GETSOCKNAME_ARG3_TYPE
#define HAVE_LIBUTIL 1
#define HAVE_IFADDRS_H 1
#define HAVE_UNISTD_H 1
#define HAVE_FNMATCH_H 1
#define HAVE_STRFTIME 1
#define HAVE_VSPRINTF 1
#define HAVE_VASPRINTF 1
#define HAVE_VSNPRINTF 1
#define HAVE_STRLCPY 1
#define HAVE_FNMATCH 1
#define HAVE_GETOPT 1
#define HAVE_GETOPT_LONG 1
#define NEED_SWAB_PROTO 1
#define HAVE_SYS_TYPES_H 1
#define HAVE_SYS_SOCKET_H 1
#define HAVE_NETINET_IN_H 1
#define HAVE_NETDB_H 1
#define HAVE_INET_NTOP 1
#define HAVE_INET_PTON 1
#define HAVE_GETNAMEINFO 1
#define HAVE_GETADDRINFO 1
#define POSIX_SIGNALS 1
#define FORTUNA 1
#define TLS_IMPL_OPENSSL 1
#define RETSIGTYPE void
#define HAVE_DIRENT_H 1
#define HAVE_DECL_STRERROR_R 1
#define HAVE_STRERROR_R 1
#define HAVE_STRDUP 1
#define HAVE_SETVBUF 1
#define HAVE_SETEUID 1
#define HAVE_SETRESUID 1
#define HAVE_SETREUID 1
#define HAVE_SETEGID 1
#define HAVE_SETRESGID 1
#define HAVE_SETREGID 1
#define HAVE_SETSID 1
#define HAVE_FLOCK 1
#define HAVE_FCHMOD 1
#define HAVE_CHMOD 1
#define HAVE_STRFTIME 1
#define HAVE_STRPTIME 1
#define HAVE_GETEUID 1
#define HAVE_SETENV 1
#define HAVE_UNSETENV 1
#define HAVE_GETENV 1
#define HAVE_GMTIME_R 1
#define HAVE_LOCALTIME_R 1
#define HAVE_MKSTEMP 1
#define HAVE_ACCESS 1
#define HAVE_GETCWD 1
#define HAVE_SRAND48 1
#define HAVE_SRAND 1
#define HAVE_SRANDOM 1
#define HAVE_STAT 1
#define HAVE_STRCHR 1
#define HAVE_STRERROR 1
#define HAVE_TIMEGM 1
#define HAVE_GETTIMEOFDAY 1
#define HAVE_UNISTD_H 1
#define HAVE_PATHS_H 1
#define HAVE_REGEX_H 1
#define HAVE_FCNTL_H 1
#define HAVE_MEMORY_H 1
#define HAVE_IFADDRS_H 1
#define HAVE_BYTESWAP_H 1
#define HAVE_ENDIAN_H 1
#define HAVE_PWD_H 1
#define HAVE_ARPA_INET_H 1
#define HAVE_ALLOCA_H 1
#define HAVE_DLFCN_H 1
#define HAVE_LIMITS_H 1
#define HAVE_STRUCT_STAT_ST_MTIMENSEC 1
#define HAVE_STRUCT_STAT_ST_MTIM_TV_NSEC 1
#define HAVE_REGEXEC 1
#define WAIT_USES_INT 1
#define USE_SIGPROCMASK 1
#define GETGROUPS_T gid_t
#define POSIX_SETJMP 1
#define SETRPCENT_TYPE void
#define ENDRPCENT_TYPE void
#define HAVE_BSWAP_16 1
#define HAVE_BSWAP_64 1
#define GETHOSTBYNAME_R_RETURNS_INT 1
#define HAVE_GETHOSTBYNAME_R 1
#define GETPWNAM_R_RETURNS_INT 1
#define HAVE_GETPWNAM_R 1
#define HAVE_GETPWUID_R 1
#define USE_DIRENT_H 1
#define POSIX_TERMIOS 1
#define POSIX_SIGTYPE 1
#define krb5_sigtype void
#define HAVE_POLL_H 1
#define HAVE_STDLIB_H 1
#define HAVE_STRING_H 1
#define HAVE_STDDEF_H 1
#define HAVE_SYS_TYPES_H 1
#define HAVE_SYS_FILE_H 1
#define HAVE_SYS_PARAM_H 1
#define HAVE_SYS_STAT_H 1
#define HAVE_SYS_TIME_H 1
#define HAVE_NETINET_IN_H 1
#define HAVE_SYS_UIO_H 1
#define HAVE_SYS_SELECT_H 1
#define HAVE_TIME_H 1
#define HAVE_PATHS_H 1
#define HAVE_ERRNO_H 1
#define TIME_WITH_SYS_TIME 1
#define SIZEOF_TIME_T 8
#define HAVE_SOCKLEN_T 1
#define HAVE_NETDB_H_H_ERRNO 1
#define HAVE_STRUCT_CMSGHDR 1
#define HAVE_STRUCT_IN_PKTINFO 1
#define HAVE_STRUCT_IN6_PKTINFO 1
#define HAVE_STRUCT_SOCKADDR_STORAGE 1
#define HAVE_SSIZE_T 1
#define HAVE_U_CHAR 1
#define HAVE_U_INT 1
#define HAVE_U_LONG 1
#define HAVE_U_INT8_T 1
#define HAVE_U_INT16_T 1
#define HAVE_U_INT32_T 1
#define HAVE_INT8_T 1
#define HAVE_INT16_T 1
#define HAVE_INT32_T 1
#define HAVE_SYS_SELECT_H 1
#define HAVE_SYS_TIME_H 1
#define HAVE_UNISTD_H 1
#define SETRPCENT_TYPE void
#define ENDRPCENT_TYPE void
#define HAVE_LIBCRYPTO 1
#define HAVE_OPENSSL_CMS 1
#define HAVE_DAEMON 1
#define HAVE_GETHOSTBYNAME_R 1
#define HAVE_GMTIME_R 1
#define HAVE_LOCALTIME_R 1
#define HAVE_READLINE 1
#define DEFCCNAME "/data/data/com.termux/files/usr/tmp/krb5cc_%{uid}"
#define DEFKTNAME "/data/data/com.termux/files/usr/etc/krb5.keytab"
#define DEFCKTNAME "/data/data/com.termux/files/usr/var/krb5/user/%{euid}/client.keytab"

The following two are missing

>#define KRB5_DNS_LOOKUP_REALM 1
#define KRB5_DNS_LOOKUP 1

**Comment by lark** [comment]
if build  with `--with-netlib=-lc`, then `--enable-dns-for-realm` is not favored and skipped.

**Comment by lark** [comment]
The build issue is easy to fix, just remove `--with-netlib=-lc`. The generated autoconf.h diff is here

--- krb5/build/include/autoconf.h	2019-02-10 18:17:16.017830334 +0800
+++ /home/builder/autoconf.h	2019-02-10 18:16:01.505487733 +0800
@@ -356,7 +356,7 @@
 /* #undef HAVE_RES_NSEARCH */

 /* Define to 1 if you have the `res_search' function */
-/* #undef HAVE_RES_SEARCH */
+#define HAVE_RES_SEARCH 1

 /* Define to 1 if you have the `re_comp' function. */
 /* #undef HAVE_RE_COMP */
@@ -586,10 +586,10 @@
 /* #undef KRB5_ATHENA_COMPAT */

 /* Define for DNS support of locating realms and KDCs */
-/* #undef KRB5_DNS_LOOKUP */
+#define KRB5_DNS_LOOKUP 1

 /* Define to enable DNS lookups of Kerberos realm names */
-/* #undef KRB5_DNS_LOOKUP_REALM */
+#define KRB5_DNS_LOOKUP_REALM 1

 /* Define if the KDC should return only vague error codes to clients */
 /* #undef KRBCONF_VAGUE_ERRORS */

But story doesn't end here. With new package built

$ KRB5_TRACE=t.log kinit random@ATHENA.MIT.EDU
kinit: Cannot find KDC for realm "ATHENA.MIT.EDU" while getting initial credentials
$ cat t.log
> [25052] 1549794759.225374: Getting initial credentials for random@ATHENA.MIT.EDU
[25052] 1549794759.225376: Sending unauthenticated request
[25052] 1549794759.225377: Sending request (178 bytes) to ATHENA.MIT.EDU
[25052] 1549794759.225378: Sending DNS URI query for _kerberos.ATHENA.MIT.EDU.
[25052] 1549794759.225379: No URI records found
[25052] 1549794759.225380: Sending DNS SRV query for _kerberos._udp.ATHENA.MIT.EDU.
[25052] 1549794759.225381: Sending DNS SRV query for _kerberos._tcp.ATHENA.MIT.EDU.
[25052] 1549794759.225382: No SRV records found
[25052] 1549794759.225383: Retrying AS request with master KDC
[25052] 1549794759.225384: Getting initial credentials for random@ATHENA.MIT.EDU
[25052] 1549794759.225386: Sending unauthenticated request
[25052] 1549794759.225387: Sending request (178 bytes) to ATHENA.MIT.EDU (master)
[25052] 1549794759.225388: Sending DNS URI query for _kerberos.ATHENA.MIT.EDU.
[25052] 1549794759.225389: No URI records found
[25052] 1549794759.225390: Sending DNS SRV query for _kerberos-master._udp.ATHENA.MIT.EDU.
[25052] 1549794759.225391: Sending DNS SRV query for _kerberos-master._tcp.ATHENA.MIT.EDU.
[25052] 1549794759.225392: No SRV records found

**Comment by lark** [comment]
Strace shows that dns resolving is tried but fails.

From the log
1. kinit reads dns server IP by reading `config properties`. But since Android 8 (API level 26/27), net.dns1, net.dns2, net.dns3, net.dns4 can't be used, so kinit gets empty result  (My phone is running Android 9)
2. uses low level resolver api (create udp socket and send dns query) and fails

> --- snip ---
socket(AF_INET, SOCK_DGRAM|SOCK_CLOEXEC, IPPROTO_IP) = 3
fchown(3, 1051, -1)                     = -1 EPERM (Operation not permitted)
openat(AT_FDCWD, "/dev/__properties__/u:object_r:config_prop:s0", O_RDONLY|O_NOFOLLOW|O_CLOEXEC) = 4
fstat(4, {st_mode=S_IFREG|0444, st_size=131072, ...}) = 0
mmap(NULL, 131072, PROT_READ, MAP_SHARED, 4, 0) = 0x70eb563000
close(4)                                = 0
bind(3, {sa_family=AF_INET, sin_port=htons(43370), sin_addr=inet_addr("0.0.0.0")}, 16) = 0
connect(3, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("0.0.0.0")}, 16) = 0
sendto(3, "\214\211\1\0\0\1\0\0\0\0\0\0\t_kerberos\6ATHENA\3MIT\3EDU\0\1\0\0\1", 42, 0, NULL, 0) = 42
pselect6(4, [3], NULL, NULL, {tv_sec=1, tv_nsec=999999479}, {NULL, 8}) = 1 (in [3], left {tv_sec=1, tv_nsec=999995833})
getsockopt(3, SOL_SOCKET, SO_ERROR, [ECONNREFUSED], [4]) = 0
close(3)                                = 0
--- snip

**Comment by Quasic** [comment]
I don't know much about this, but the switch used to be ```--with-netlib=-lcares``` which may be the same thing. I think it was added because the default didn't work. Idk if there is another setting for that switch that would suit your needs, but @vishalbiswas would probably know better.

**Comment by lark** [comment]
> I don't know much about this, but the switch used to be `--with-netlib=-lcares` which may be the same thing. I think it was added because the default didn't work. Idk if there is another setting for that switch that would suit your needs, but @vishalbiswas would probably know better.

When the package first added, it's
`TERMUX_PKG_EXTRA_CONFIGURE_ARGS=" --with-readline --with-netlib=-lcares --without-system-verto"`

later, it's changed to
`TERMUX_PKG_EXTRA_CONFIGURE_ARGS="--with-readline
--without-system-verto
--with-netlib=-lc
--enable-dns-for-realm
--sbindir=$TERMUX_PREFIX/bin
--with-size-optimizations
--with-system-db
"`

My guess is nobody really tested  it  at that time because no one suggested simple test case.

I did some investigation to confirm that mutual exclusion of `--with-netlib` and `--enable-dns-for-realm` was not introduced later without notice. But autotools is overwhelming for me and I am not sure.

**Comment by quentinmit** [comment]
This is still a problem with the current version `1.19.2-1` of `krb5`. It sounds like @Quasic has a proposed solution above, no?

**Comment by xtkoba** [comment]
Seems like basically the same as #1225. Should be fixed by linking against `libresolv-wrapper` (available in PR #8064, not merged yet).

**Comment by xtkoba** [comment]
Is this a correct behavior?

```
$ kinit randomuser@ATHENA.MIT.EDU
kinit: Client 'randomuser@ATHENA.MIT.EDU' not found in Kerberos database while getting initial credentials
```

**Comment by quentinmit** [comment]
> Is this a correct behavior?
> 
> ```
> $ kinit randomuser@ATHENA.MIT.EDU
> kinit: Client 'randomuser@ATHENA.MIT.EDU' not found in Kerberos database while getting initial credentials
> ```

Yes, that is the correct response. `kinit moira@ATHENA.MIT.EDU` should likewise prompt for a password.

**Comment by xtkoba** [comment]
Thanks for the response.

```
$ kinit moira@ATHENA.MIT.EDU
Password for moira@ATHENA.MIT.EDU:
```

So I believe this is fixed by PR #8160.

**Comment by quentinmit** [comment]
That is awesome! Thanks!
